🎉 New implementation for AWS Secret Manager for issue

mauricioalarcon opened the pull request:
## What
Initial implementation for AWS Secret Manager for [issue](https://github.com/airbytehq/airbyte/issues/10518)
https://github.com/airbytehq/airbyte/issues/10518

Added new implementation `AWSSecretManagerPersistence` and integration tests `AWSSecretManagerPersistenceIntegrationTest`

## How
A new implementation of `SecretPersistence` to support AWS Secret Manager `AWSSecretManagerPersistence` 
New Integration tests as suggested on the open Issue, similar to GCP secret manager

![image](https://user-images.githubusercontent.com/558874/203184092-d96d7ac3-92e0-4310-8ac8-1f96122dccd2.png)


## 🚨 User Impact 🚨
No breaking changes, you need a new set of AWS credentials and AWS cli installed in order to run integration test.

Also note that the credentials need the following
- "secretsmanager:GetResourcePolicy",
- "secretsmanager:GetSecretValue",
- "secretsmanager:DescribeSecret",
- "secretsmanager:ListSecretVersionIds"
- "secretsmanager:DeleteSecret"

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Feature Implemenation</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)



## Tests

<details><summary><strong>Unit</strong></summary>

*Building Airbyte Sub Build: PLATFORM
Type-safe dependency accessors is an incubating feature.

> Configure project :
configuring docker task for airbyte-bootloader
configuring docker task for airbyte-connector-builder-server
configuring docker task for airbyte-container-orchestrator
configuring docker task for airbyte-cron
configuring docker task for airbyte-proxy
configuring docker task for airbyte-server
configuring docker task for airbyte-temporal
configuring docker task for airbyte-webapp
configuring docker task for airbyte-workers
configuring docker task for init
configuring docker task for db-lib
configuring docker task for reporter

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.4/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 1s
13 actionable tasks: 13 up-to-date
*

</details>

<details><summary><strong>Integration</strong></summary>


*Gist with results [here](https://gist.github.com/mauricioalarcon/5af2b4be7959697db52728bb8bc30b92)*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Gist with results [here](https://gist.github.com/mauricioalarcon/50424bf3acffcacf2c9a811742a42fc9)*

</details>

mauricioalarcon committed:
🎉 New implementation for AWS Secret Manager for issue
#10518

Added new implementation AWSSecretManagerPersistence and integration tests AWSSecretManagerPersistenceIntegrationTest

A new implementation of SecretPersistence to support AWS Secret Manager AWSSecretManagerPersistence
New Integration tests as suggested on the open Issue, similar to GCP secret manager

marcosmarxm commented:
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

mauricioalarcon commented:
Hi @marcosmarxm, thanks for your reply; I understand Thanksgiving is a slow week for all of us. I'm looking forward to any feedback you guys have on this.

In the meantime, I've been merging from `master` any new changes to my branch.

Happy Thanksgiving 

 Mauricio

davinchia commented:
Thanks for the contribution @mauricioalarcon !

@pmossman will take a look over the next few days!

mauricioalarcon commented:
> Thanks for the contribution @mauricioalarcon !
> 
> @pmossman will take a look over the next few days!

Thank you @davinchia - I'm looking forward to hear from @pmossman

pmossman commented:
Hey @mauricioalarcon , thanks for the contribution! This is on my todo-list for this week. At first glance, it looks good to me, but I'll want to spend some time trying it out before I approve it. I'll keep you updated!

pmossman commented:
Actually @mauricioalarcon it looks like our Snyk security scan found a vulnerability introduced by that aws-secretsmanager package that your PR introduces:

<img width="658" alt="image" src="https://user-images.githubusercontent.com/7798112/204937701-4b78d8a7-0ef6-4687-b9f3-fa4c6802d922.png">

Perhaps there's a newer version of the package with that vulnerability fixed? Or perhaps there's a way to force one of the fixed versions of jackson-databind as indicated in the screenshot?

mauricioalarcon committed:
Updated version to `aws-secretsmanager-caching-java:1.0.2`
- Added new argument check for AWS Secret Manager.

pmossman commented:
Hmm that's odd, Snyk is now reporting a different security issue, this time for `jackson-databind:2.12.3` even though it looks like gradle is ultimately using 2.13.3

I might need to pull in some extra eyes on this one for debugging gradle dependencies, I'll try to check back in tomorrow, sorry for the trouble @mauricioalarcon !

Commit:
exclude jackson-databind from aws dependency and run gradle format

Commit:
bump aws-related dependencies to prevent old version of jackson-databind from being pulled in

pmossman commented:
@mauricioalarcon I've determined that Snyk's complaint about the dependency is a false alarm, since we declare an updated version of jackson-databind globally, but Snyk isn't able to detect it. I've manually marked your PR as resolved in Snyk so we shouldn't see that check fail any more.

I'm currently looking into how our CI system will run the integration test for the AWS secret manager. We have some specialized setup for testing our GCP secret manager in our CI, so I'm going to see what we need to do to make the AWS secret manager work in a similar fashion.

It might be non-trivial, since I think we have a dedicated GCP project stood up for integration tests, and we might need to create something similar for AWS. 

I'll keep working to move this forward, just wanted to give you a quick end-of-week update. Have a good weekend!

mauricioalarcon commented:
Thanks @pmossman, for the update and work on this. The only thing needed for the integration test will be just an account where we can run these.
Also, if you feel, I think I can put together a unit test that makes uses of mockito and assert the same functionality go along or avoid the integration test if that gets to complex.

lmk if I can help in any way, shape or form.

pmossman commented on airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/split_secrets/AWSSecretManagerPersistence.java:
```diff
@@ -0,0 +1,118 @@
+   * @see SecretCache
+   */
+  public AWSSecretManagerPersistence() {
+    this.client = AWSSecretsManagerClientBuilder.defaultClient();
```
@mauricioalarcon before we get this PR merged, can you incorporate the AWS Secret Manager into https://github.com/airbytehq/airbyte/blob/92f71accea6674f04cabf1bcdc8a48f7f4866905/airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretPersistence.java similar to `GOOGLE_SECRET_MANAGER` and `VAULT`? 

We have two "modes" for secret persistence layers, "ephemeral" and "long-lived". For Google Secret Manager, we define separate TTLs for each, but for Vault, both use the same underlying implementation. I think for now, we can use the same implementation of the AWS Secret Manager for both, unless you want to take a look at the Google internals and replicate something similar for AWS, by all means go for it!

One final important note: these other secret persistence classes use credentials and configurations from our `Configs.java`, and I think it'd make sense to add some new configs for AWS Secrets Manager so that we're not relying on any defaults.

For example, these configs power the Vault-based secret persistence: https://github.com/airbytehq/airbyte/blob/master/airbyte-config/config-models/src/main/java/io/airbyte/config/Configs.java#L134-L150

And the implementation for these methods are here: https://github.com/airbytehq/airbyte/blob/master/airbyte-config/config-models/src/main/java/io/airbyte/config/EnvConfigs.java#L396-L409

This will make it easier for various environments to configure the AWS Secret manager, without needing to rely on default profiles, which can get particularly cumbersome in some environments where multiple profiles exist.

Let me know if you have any questions around any of that! Really appreciate your contribution here :)

mauricioalarcon replied:
Thanks @pmossman all of these changes make sense. I'll work on these and send an update to the PR.

mauricioalarcon replied:
@pmossman I adjusted the PR with the requested changes; I added the new config keys as part of the instructions at `docs/operator-guides/configuring-airbyte.md`, I'm not sure where else I need to update, if any.

As for the "ephemeral", that's not an option for AWS; that will require a bit more to mimic the TTL. I'll research a bit more on this, and I'll get back to you.

Thank you

Commit:
- Added new entry for `AWS_SECRET_MANAGER` into SecretPersistence.java
- Externalized configurations for AWS Secret Manager into Configs and EnvConfigs.java as requested.
- Adjusted test to use new constructor for `AWSSecretManagerPersistence`
- Augmented instructions configuring-airbyte.md with the new config keys.

Commit:
Formatting code.

mauricioalarcon commented:
@pmossman - I think I know why the deployments were failing; due to [formatting](https://github.com/airbytehq/airbyte/actions/runs/3659153968/jobs/6184858803) ?  sorry, - I've just ran formatter and updated the PR.

pmossman commented:
@mauricioalarcon I think this is very close, I'm just looking into a failing acceptance test that I believe is unrelated to your change. I think the s3 dependency version bump that I included in this PR is causing an S3-related connector to fail in our test suite, so I'm going to revert it in a separate branch and see what happens to our tests. Your changes in this PR look great so once I figure out this final dependency issue we should be good to merge

Commit:
revert dependency version bump

pmossman reviewed (approved):
@mauricioalarcon I verified that your PR passes our full suite of acceptance tests on another branch of mine, and I verified that the security warning from Snyk can be disregarded as it isn't smart enough to see that we're pulling in a fixed version of transitive dependencies.

Thanks so much for your contribution and for making the AWS Secret Manager compatible with our EnvConfigs setup! Sorry this took so much back and forth, we're still working on making our tooling around community PRs more streamlined, and your particular PR gave us some great data points for improving the process.

Thanks again!

mauricioalarcon commented:
@pmossman That's amazing - thanks a lot; I'm happy to help with such a fantastic product that's Airbyte.

Gauravshah commented on airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretPersistence.java:
```diff
@@ -34,6 +34,9 @@ static Optional<SecretPersistence> getLongLived(final @Nullable DSLContext dslCo
       case VAULT -> {
         return Optional.of(new VaultSecretPersistence(configs.getVaultAddress(), configs.getVaultPrefix(), configs.getVaultToken()));
       }
+      case AWS_SECRET_MANAGER -> {
+        return Optional.of(new AWSSecretManagerPersistence(configs.getAwsAccessKey(), configs.getAwsSecretAccessKey()));
+      }
```
any reason to not support IAM role based authentication ?

kjoconnor replied:
The Airbyte project at large doesn't seem to support role based auth at all, which is pretty rough.

madianas21 commented:
Are there any docs on how to fetch secrets from `AWS Secret Manager` ?
